pkg/serial/handshake.go

```diff
@@ -162,9 +162,26 @@ func HandshakeClient(ctx context.Context, conn net.Conn, debug bool) error {
 	return nil
 }
 
-func readMultiple(conn net.Conn, b []byte) (int, error) {
+func readMultiple(ctx context.Context, conn net.Conn, b []byte) (int, error) {
 	if runtime.GOOS != "windows" {
-		return conn.Read(b)
+		type ioret struct {
+			n   int
+			err error
+		}
+		c := make(chan ioret, 1)
+
+		go func() {
+			n, err := conn.Read(b)
+			c <- ioret{n, err}
+			close(c)
+		}()
+
+		select {
+		case ret := <-c:
+			return ret.n, ret.err
+		case <-ctx.Done():
+			return 0, ctx.Err()
```

Comment:
Implementing rawconn.SetReadDeadline might be easier to parse and is consistent with expected net.Conn behaviour.
I'd actually prefer conn.SetReadDeadline(ctx) but it doesn't match the interface